File: 0238-product-of-array-except-self/0238-product-of-array-except-self.go
/*
func productExceptSelf(nums []int) []int {
    forwards := make([]int, len(nums))
    forwards[0] = nums[0]
    for i := 1; i < len(forwards); i++ {
        forwards[i] = nums[i] * forwards[i-1]
    }
    
    backwards := make([]int, len(nums))
    backwards[len(backwards)-1] = nums[len(nums)-1]
    for i := len(backwards)-2; i >= 0; i-- {
        backwards[i] = nums[i] * backwards[i+1]
    }
    
    result := make([]int, len(nums))
    for i := 1; i < len(result)-1; i++ {
        result[i] = forwards[i-1] * backwards[i+1]
    }
    result[0] = backwards[1]
    result[len(result)-1] = forwards[len(forwards)-2]
    
    return result
}
*/
func productExceptSelf(nums []int) []int {
    products := make([]int, len(nums))
    pbefore := 1
    for i := 0; i < len(products); i++ {
        products[i] = pbefore
        pbefore *= nums[i]
    }
    
    pafter := 1
    for i := len(nums)-1; i >= 0; i-- {
        products[i] *= pafter
        pafter *= nums[i]
    }
    return products
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* productExceptSelf(int* nums, int numsSize, int* returnSize){
    int product, i;
    int forwards[numsSize];
    int backwards[numsSize];
    int *result = malloc(sizeof(int) * numsSize);
    
    forwards[0] = nums[0];
    for (i = 1; i < numsSize; i++) {
        forwards[i] = forwards[i-1] * nums[i];
    }
    
    backwards[numsSize-1] = nums[numsSize-1];
    for (i = numsSize-2; i >= 0; i--) {
        backwards[i] = backwards[i+1] * nums[i];
    }
    
    result[0] = backwards[1];
    result[numsSize-1] = forwards[numsSize-2];
    for (i = 1; i < numsSize-1; i++)
        result[i] = forwards[i-1] * backwards[i+1];
    
    *returnSize = numsSize;
    return result;
}